### ISSUE TITLE
OpenType features on decimal separators are disabled

### ISSUE THREAD

**Issue by denis-sokolov** [atom/language-javascript#364]
This is how a number renders:
```html
<span class="constant numeric decimal js">
  42<span class="meta delimiter decimal period js">.</span>2
</span>
```

Because the decimal is in a separate element, OpenType contextual replacement feature does not work across the entire number. For instance, a typeface might contain two different period glyphs – one for text, and the other for decimal separator.

This is how the DOM needs to look for the feature to work:
```html
<span class="constant numeric decimal js">
  42.2
</span>
```

I understand that somebody might be relying on this extra element for, perhaps, syntax highlighting the decimal separator. There seems to be a clear conflict of features, and it is up to you to make the call.

**Comment by MaximSokolov** [comment]
Do you mean ligatures? Could you post a screenshot and sign what font on it? I curious how does it look like.
> I understand that somebody might be relying on this extra element for, perhaps, syntax highlighting the decimal separator.

Originally it was part of fixing numbers like `.1` (#319) Nobody asked for it before.

**Comment by Alhadis** [comment]
@denis-sokolov I sorely doubt OpenType ligatures are used in raw source code. I can't for the life of me imagine why anybody would want the decimal point of a float displayed differently to regular accessors.

**Comment by MaximSokolov** [comment]
> I sorely doubt OpenType ligatures are used in raw source code

e.g. [FiraCode](https://github.com/tonsky/FiraCode) and [Hasklig](https://github.com/i-tu/Hasklig). Though I am not sure how this affects decimal point

**Comment by Alhadis** [comment]
Sorry, I should've clarified what I meant. Fixed.

**Comment by Alhadis** [comment]
Paging @denis-sokolov one last time, because I'm curious to find out what font he's using that actually uses this sort of ligature.

**Comment by denis-sokolov** [comment]
Admittedly, it's not a popular typeface. Purisa-code, it's on my GitHub
profile.

I believe a decimal number is a single token: we don't highlight different
digits in a non-decimal number: 6744. What's different about the decimal
part? In my typeface the decimal point is tiny and you read the entire
number at once.

But the dot as a property accessor (or method invocation) is majorly
important in the code. In code chains it is important to see the steps
separated: `foo.bars.filter().forEach()`.

**Comment by Alhadis** [comment]
To be honest, I agree with the part about numbers being perceived as a solitary unit. Similarly, a leading sign indicator shouldn't be highlighted differently, either:

<img src="https://cloud.githubusercontent.com/assets/2346707/18026121/b954f78e-6c80-11e6-9c18-bf38c8009a9c.png" width="225" alt="Figure 1" />

Oddly enough, the sign indicator of an exponential component doesn't receive different highlighting, which makes the overall composition pretty inconsistent.

@50Wliu, thoughts?

**Comment by 50Wliu** [comment]
It has been previously attempted to include the optional leading `+/-` as part of the number.  That didn't work because `- 5` in `3 - 5` would get tokenized as a number and not as an operator and a number.

Remember, all this is customizable in your themes.

**Comment by denis-sokolov** [comment]
But isn't the goal of a grammar to understand the language and distinguish
unary and binary operators?

**Comment by 50Wliu** [comment]
Yes, however there are limitations to the regex-based model.  Take, for example, this JavaScript snippet:
```js
let x = 3
-5;
```

How does the parser know that the `-` is actually an operator?

**Comment by denis-sokolov** [comment]
I am not one to comment to the viability of the current model or why it's been chosen. But I don't think we actually need 100% accuracy, do we? Also, this counter-example does not apply to the decimal point that the issue is about. :)

**Comment by Alhadis** [comment]
Considering this grammar is being used to highlight the most popular programming language on Earth, on the biggest open source community in the world (all of GitHub), I'd say accuracy is pretty damn important.

**Comment by denis-sokolov** [comment]
That's an abstract statement. The context where I said we don't need 100% accuracy is where currently we arguably have an even worse accuracy (detecting unary vs binary `-`). Perfect is the enemy of good.

**Comment by 50Wliu** [comment]
Back to the decimal point...there may very well be people who would prefer that it was stylized differently than the surrounding numbers, as you said in the original post.  In fact, the syntax theme I'm using colors it differently.

**Comment by Alhadis** [comment]
I think this is a non-issue, quite frankly.

**Comment by denis-sokolov** [comment]
@50Wliu, as I said, it's your call. In my personal opinion it is as strange to style the decimal point differently as styling individual digits. The entire number is a number, and I would have preferred for it to be rendered as a single number. I don't have any further argument.

**Comment by esdoppio** [comment]
Ligatures set aside, I say we comply with the parsers when we can.

Resulted tokens parsing `-2.2e-2; 4.2;`:

```
Punctuator(-)
Numeric(2.2e-2)
Punctuator(;)
Numeric(4.2)
Punctuator(;)
```

![e563587c-3df8-4942-81b5-4d2a64a24f0f](https://cloud.githubusercontent.com/assets/18070858/18228749/65d50838-728e-11e6-8d98-350aa75b03a2.png)

**Comment by 50Wliu** [comment]
I'm thinking of flip-flopping on this issue based on my work with language-java.  Give me a few more days to think about this.

**Comment by Alhadis** [comment]
\*googles *flip-flopping* to figure out what that meant\*

<img width="330" alt="screen shot 2016-12-08 at 6 11 02 pm" src="https://cloud.githubusercontent.com/assets/2346707/21000951/c1c9feae-bd71-11e6-9044-5d63d30de15c.png">

Strewth.
